ccm_web/server/src/canvas/invalid.token.interceptor.ts

```diff
@@ -0,0 +1,36 @@
+import { Request } from 'express'
+import { CallHandler, ExecutionContext, HttpException, HttpStatus, Injectable, NestInterceptor, UnauthorizedException } from '@nestjs/common'
+import { Observable } from 'rxjs'
+import { catchError } from 'rxjs/operators'
+
+import { CanvasService } from './canvas.service'
+import { RequestWithoutUserError } from '../user/user.errors'
+import { InvalidTokenRefreshError } from './canvas.errors'
+
+@Injectable()
+export class InvalidTokenInterceptor implements NestInterceptor {
+  constructor (private readonly canvasService: CanvasService) {}
+
+  async intercept (context: ExecutionContext, next: CallHandler): Promise<Observable<unknown>> {
+    const req = context.switchToHttp().getRequest<Request>()
+    const user = req.user
+    if (user === undefined) throw new RequestWithoutUserError()
+
+    return next.handle().pipe(
+      catchError(async (err) => {
+        // Other unauthorized cases related to JWT or sessions will be caught by guards.
+        if (
+          err instanceof InvalidTokenRefreshError ||
+          (err instanceof HttpException && err.getStatus() === HttpStatus.UNAUTHORIZED)
+        ) {
+          await this.canvasService.deleteTokenForUser(user)
```

Comment:
There seems to be some portion of programmers who espouse that exceptions shouldn't be used for controlling the flow of execution. I guess this could be considered an instance of that, and we could move the use of `canvasService.deleteTokenForUser` to `canvasService.refreshToken` and a handful of places in `APIService`. I was trying to keep the code DRY, but I can see the argument that the interceptor usage makes the code more difficult to read.